tools/cache_creator/cache_creator.cpp

```diff
@@ -161,6 +162,82 @@ bool hexStringToUuid(llvm::StringRef hexStr, llvm::MutableArrayRef<uint8_t> outU
   return true;
 }
 
+static llvm::Error createBadPalMetadataError() {
+  return llvm::createStringError(llvm::object::object_error::invalid_file_type, "Unexpected PAL metadata format");
+}
+
+static llvm::Expected<llvm::msgpack::MapDocNode> get128BitCacheHashNode(llvm::msgpack::Document &document) {
+  llvm::msgpack::DocNode node = document.getRoot();
+  if (!node.isMap())
+    return createBadPalMetadataError();
+  node = node.getMap(false)["amdpal.pipelines"];
+  if (!node.isArray())
+    return createBadPalMetadataError();
+  node = node.getArray(false)[0];
+  if (!node.isMap())
+    return createBadPalMetadataError();
+  node = node.getMap(false)[".xgl_cache_info"];
+  if (!node.isMap())
+    return createBadPalMetadataError();
+  return node.getMap(false);
+}
+
+static llvm::Expected<Util::MetroHash::Hash> extractCacheHash(llvm::msgpack::MapDocNode &cacheInfoNode) {
+  llvm::msgpack::DocNode node = cacheInfoNode[".128_bit_cache_hash"];
+  if (!node.isArray())
+    return createBadPalMetadataError();
+  llvm::msgpack::ArrayDocNode hashNode = node.getArray(false);
+  if (hashNode[0].getKind() != llvm::msgpack::Type::UInt || hashNode[1].getKind() != llvm::msgpack::Type::UInt)
+    return createBadPalMetadataError();
+  Util::MetroHash::Hash hash = {};
+  hash.qwords[0] = hashNode[0].getUInt();
+  hash.qwords[1] = hashNode[1].getUInt();
+  return hash;
+}
+
+static llvm::Expected<llvm::VersionTuple> extractLlpcVersion(llvm::msgpack::MapDocNode &cacheInfoNode) {
+  llvm::msgpack::DocNode node = cacheInfoNode[".llpc_version"];
+  if (!node.isString())
+    return createBadPalMetadataError();
+  llvm::VersionTuple versionTuple;
+  if (versionTuple.tryParse(node.getString()))
+    return createBadPalMetadataError();
+  if (!versionTuple.getMinor().hasValue())
+    return createBadPalMetadataError();
+  return versionTuple;
+}
+
+// =====================================================================================================================
+// Tries to extract cache hash and LLPC version from a PAL metadata note blob.
+//
+// @param noteBlob : The input metadata blob
+// @returns : The cache info on success, or error on failure
+llvm::Expected<ElfLlpcCacheInfo> getCacheInfoFromMetadataBlob(llvm::StringRef noteBlob) {
+  llvm::msgpack::Document document;
+  if (!document.readFromBlob(noteBlob, false))
+    return createBadPalMetadataError();
+
+  // Get XGL cache info node from document.
```

Comment:
This comment and the two comments below are not needed because you are essentially repeating the function name.